enhancements/baremetal/an-slo-for-baremetal.md

```diff
@@ -0,0 +1,592 @@
+namespace from the MAO to the CBO.
+
+We assume that the CBO will first appear in `4.N.0` and take over from
+the MAO in that release. We can merge changes into `4.N-1.y` paving
+the way for this and, if necessary, require an upgrade to given
+`4.N-1.y` version before the upgrade to `4.N.0`. Since we only support
+upgrading directly from `4.N-1` to `4.N`, we can remove all upgrade
+handling code in `4.N+1`.
+
+#### Prior Art
+
+We will take inspiration from the
+[Separate OAuth API-Resources](https://github.com/openshift/enhancements/blob/master/enhancements/authentication/separate-oauth-resources.md)
+enhancement
+
+> Add code in 4.(n-1) to the cluster-openshift-apiserver-operator that
+> prevents it from managing an apiservice if it is "claimed" by the
+> cluster-authentication-operator via an annotation and if the
+> cluster-authentication-operator is at least at 4.n.
+
+In the related
+[cluster-openshift-apiserver-operator PR](https://github.com/openshift/cluster-openshift-apiserver-operator/pull/294)
+the details are clear - `cluster-authentication-operator` indicates
+it is "at least at 4.n" by setting the `managingOAuthAPIServer` to
+`True` in the status field of its `authentication` resource. Then
+`cluster-authentication-operator` "claims" an `apiservice` by setting
+the `authentication.operator.openshift.io/managed` annotation.
+
+Presumably the "at least at 4.n" check is required in addition to the
+annotation to handle a downgrade scenario - where the annotation is
+set, but `cluster-authentication-operator` has been downgraded.
+
+#### Design
+
+We need the following changes to MAO:
+
+1. Respect a "gate" allowing CBO to "claim" the resource.
+
+   We will use [the existing `machine.openshift.io/owned` annotation](http://github.com/openshift/machine-api-operator/pull/424)
+   on the `metal3` deployment to indicate that MAO is managing the
+   resource. A new `baremetal.openshift.io/owned` annotation will
+   indicate that CBO is in control. Only one of these annotations
+   should be set on a resource but, for the avoidance of doubt, CBO
+   takes precedence.
+
+   Before CBO is be added to the `4.N-dev` release image, MAO will
+   need to respect this gate - it must not attempt to manage the
+   `metal3` deployment if CBO has claimed the resource. This MAO
+   change will be backported to `4.N-1`.
+
+2. An "at least at 4.N" downgrade check.
+
+   In order to guard against a downgrade scenario where CBO had
+   claimed the `metal3` deployment - but the cluster has since been
+   downgraded and CBO (manually) removed - we need some way of
+   checking CBO is actually still around. We will use the existence of
+   a `baremetal` clusteroperator for this purpose. Users would need to
+   manually remove this resource after a downgrade.
+
+3. A change to remove `metal3` deployment from MAO.
+
+   In `4.N-dev`, once the CBO has been added to the release, we can
+   remove all awareness of the `metal3` deployment from MAO.
```

Comment:
I'm not sure why all this is needed. why do we need to do any check for the CBO to be available or annotations at all?
The CBO either is running or not. If it is running in 4.N, then that version of MAO needs to know nothing about baremetal. During a 4.N upgrade the CBO just takes over, in a downgrade the old mao code takes over.

Reply:
I think the main issue is that during an upgrade, we can't actually guaranteee that the old MAO is shut down before the new CBO is running.

If you can explain how we could do that, then great - nobody involved loves this extra complexity.

Downgrade is perhaps a corner case not worth over-designing for, but ... AFAICT, in a downgrade scenario, the CBO will not be automatically deleted by the CVO. So you'll have metal3-managing MAO and CBO running, until the admin manually deletes the CBO.

Also if you could explain why this was needed for the oauth example (in the "prior art" section) but not needed in this case, then that would be super helpful.

Reply:
Also, you at least have a very short term challenge with this transition - you can't merge "remove metal3 management from MAO" and "add CBO to the release image" as an atomic change. There will be some set of release images with either a regression or a conflict.

(But again, my main concern is e.g. a 4.6->4.7 upgrade where we have to transition from MAO to CBO, and we can't guarantee that the 4.6 MAO is shut down first)

Reply:
Thanks @markmc I shared all those concerns. I think we could possibly take advantage of prefix order https://github.com/openshift/cluster-version-operator/blob/master/docs/dev/upgrades.md#upgrades-and-order for upgrades to guarantee that the old MAO is shut down before the new CBO is running.

For the downgrade corner case hopefully let the mao check for the CBO existence should be enough, so we can remove some complexity without getting annotations involved.

fwiw the rationale for the annotation in mao was purely to give us control over which specific deployments the informer is watching in that namespace, there's probably more elegant ways to configure the informer to do this.

Reply:
> Thanks @markmc I share all those concerns. I think we could possibly take advantage of prefix order https://github.com/openshift/cluster-version-operator/blob/master/docs/dev/upgrades.md#upgrades-and-order for upgrades to guarantee that the old MAO is shut down before the new CBO is running.

(duplicating from openshift/machine-api-operator#689)

Yeah, but it's pretty ugly to put CBO in a later runlevel just for this temporary upgrade handover. Even that link suggests using another mechanism - "Over time, operators should be better at detecting their prerequisites without overcomplicating or risking the predictability of upgrades"

> For the downgrade corner case hopefully let the mao check for the CBO existence should be enough, so we can remove some complexity without getting annotations involved.

I think I could buy that the "existence of a baremetal clusteroperator" check could serve the same purpose as the annotation - at least I'm not seeing any obvious reason why not, months after writing the enhancement! (i.e. if there's a good reason, I should have captured it in the enhancement)

> fwiw the rationale for the annotation in mao was purely to give us control over which specific deployments the informer is watching in that namespace, there's probably more elegant ways to configure the informer to do this.

Just to be open about some of the irritation here - the level of scrutiny this minor, temporary, and exhaustively documented piece of "inelegance" is getting seems way out of proportion with e.g. openshift/machine-api-operator#424. Particularly since this entire move is motivated by the clear desire for the MAO maintainers to not have to worry about bare metal use cases.

Reply:
Yeh we are totally on the same page. I meant is not elegant how is done in mao, so hopefully we could avoid perpetuate the pattern and later improve it in mao. The proposal and https://github.com/openshift/machine-api-operator/pull/689 looks all good to me in its current form. I was just making sure I got the rationale behind all details after being away for a little while. Thanks for the clarifications.